### PR TITLE
Implement rust_pump_stream (4.2.1): Rust, shim, tests, docs

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Implement rust_pump_stream as a PyO3-exported function with GIL release, a default 64 KiB buffer, and proper error propagation. Add unit and behavioural tests, a Python shim cuprum/_streams_rs.py, and comprehensive documentation updates. Update the roadmap to mark 4.2.1 as done.

### Rust
- Add rust_pump_stream(reader_fd: int, writer_fd: int, buffer_size: int = 65536) -> int
- Validate buffer_size > 0 and raise ValueError on invalid input
- Release the GIL for the entire read/write loop using Python::allow_threads
- Use non-owning file descriptor wrappers so Python FDs are not closed
- On BrokenPipe/ConnectionReset during writes, stop writing but continue draining reads (consistent with _pump_stream())
- Map std::io::Error to OSError and propagate to Python
- Return total bytes transferred

### Python shim
- Introduce optional cuprum/_streams_rs.py shim to re-export rust_pump_stream from the native module when needed, keeping public APIs stable (cuprum._rust_backend_native remains the primary native module)

### Tests
- New unit tests: cuprum/unittests/test_rust_streams.py
  - Normal transfer with default buffer
  - Custom buffer_size behavior
  - Invalid/closed file descriptors raise OSError
  - Broken pipe scenario: downstream closes early; continue draining input without raising
- Behavioural tests (pytest-bdd): tests/behaviour/test_rust_streams_behaviour.py
  - Skips cleanly if the native extension is unavailable
  - Verifies data written to source pipe is observed at destination after rust_pump_stream()
- Feature file: tests/features/rust_streams.feature
  - Scenario: Rust pump stream transfers data between pipes

### Documentation
- New ExecPlan document: docs/execplans/4-2-1-rust-pump-stream.md
- Update design/docs: docs/cuprum-design.md (Section 13) to reflect the finalized Rust module name, function signature, default buffer, and error semantics
- User guide: docs/users-guide.md notes on Rust pump stream existence and consumer-visible behavior
- Roadmap: docs/roadmap.md mark 4.2.1 as done when implemented and tested

### Artifacts
- Rust implementation: rust/cuprum-rust/src/lib.rs (rust_pump_stream)
- Optional Python shim: cuprum/_streams_rs.py
- Tests:
  - cuprum/unittests/test_rust_streams.py
  - tests/behaviour/test_rust_streams_behaviour.py
  - tests/features/rust_streams.feature
- Documentation:
  - docs/execplans/4-2-1-rust-pump-stream.md
  - docs/cuprum-design.md (Section 13)
  - docs/users-guide.md
  - docs/roadmap.md

## Rationale
This change establishes a robust, Rust-backed data-pump that operates outside the GIL, with deterministic buffering and well-mapped Python exceptions. The design aligns with existing Python I/O semantics, ensuring predictable behavior under normal and error paths, and provides thorough test coverage and documentation for maintainability.

## Notes
- If the native module name or API boundary needs adjustment, updates should be reflected in Stage A/B decisions and corresponding docs.
- Behavioural tests are designed to skip gracefully if the native extension is not available in the test environment.

📎 **Task**: https://www.devboxer.com/task/790618ac-0f24-41b2-b5f8-cf88c8c6078d

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust/cuprum-rust/src/lib.rs](https://github.com/leynos/cuprum/pull/29/files/53bf6d9ecdd83d360e649d00a0945b47583e429e#diff-4f2be807475310e1b24be797a92fb6885a48461151ead5b9c18f93070962399a)

Comment on lines +114 to +154

~~~rust
fn pump_stream_files(
    reader: &mut File,
    writer: &mut File,
    buffer_size: usize,
) -> Result<u64, io::Error> {
    let mut buffer = vec![0_u8; buffer_size];
    let mut total_written = 0_u64;
    let mut writer_open = true;

    loop {
        let read_len = reader.read(&mut buffer)?;
        if read_len == 0 {
            break;
        }
        if writer_open {
            let chunk = buffer.get(..read_len).ok_or_else(|| {
                io::Error::new(
                    io::ErrorKind::InvalidData,
                    "read length exceeds buffer size",
                )
            })?;
            if let Err(err) = writer.write_all(chunk) {
                if matches!(
                    err.kind(),
                    io::ErrorKind::BrokenPipe | io::ErrorKind::ConnectionReset
                ) {
                    writer_open = false;
                } else {
                    return Err(err);
                }
            } else {
                let read_len_u64 = u64::try_from(read_len).map_err(|_| {
                    io::Error::new(io::ErrorKind::Other, "read length overflow")
                })?;
                total_written = total_written.saturating_add(read_len_u64);
            }
        }
    }

    Ok(total_written)
}
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=29&biomarker=Bumpy+Road+Ahead&filename=rust%2Fcuprum-rust%2Fsrc%2Flib.rs&method=pump_stream_files)
pump_stream_files has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[cuprum/unittests/test_rust_streams.py](https://github.com/leynos/cuprum/pull/29/files/53bf6d9ecdd83d360e649d00a0945b47583e429e#diff-3708370c7648b8a254b1f926a193dc609f13fb6f795978eae585e97ad79745d0)

Comment on lines +43 to +65

~~~python
def test_rust_pump_stream_transfers_bytes() -> None:
    """rust_pump_stream transfers bytes between pipes."""
    streams = _load_streams_rs()
    payload = b"cuprum-stream-payload"
    in_read, in_write = os.pipe()
    out_read, out_write = os.pipe()

    try:
        os.write(in_write, payload)
        _safe_close(in_write)

        transferred = streams.rust_pump_stream(in_read, out_write)

        _safe_close(out_write)
        output = _read_all(out_read)
    finally:
        _safe_close(in_read)
        _safe_close(in_write)
        _safe_close(out_read)
        _safe_close(out_write)

    assert output == payload
    assert transferred == len(payload)
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=29&biomarker=Code+Duplication&filename=cuprum%2Funittests%2Ftest_rust_streams.py&method=)
The module contains 2 functions with similar structure: test_rust_pump_stream_respects_buffer_size,test_rust_pump_stream_transfers_bytes

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
cuprum/unittests/test_rust_streams.py: 

### What lead to degradation?
The module contains 2 functions with similar structure: test_rust_pump_stream_respects_buffer_size,test_rust_pump_stream_transfers_bytes

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Bumpy Road Ahead
rust/cuprum-rust/src/lib.rs: pump_stream_files

### What lead to degradation?
pump_stream_files has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

### Why does this problem occur?
A Bumpy Road is a function that contains multiple chunks of nested conditional logic inside the same function. The deeper the nesting and the more bumps, the lower the code health.
A bumpy code road represents a lack of encapsulation which becomes an obstacle to comprehension. In imperative languages there’s also an increased risk for feature entanglement, which leads to complex state management. CodeScene considers the following rules for the code health impact: 1) The deeper the nested conditional logic of each bump, the higher the tax on our working memory. 2) The more bumps inside a function, the more expensive it is to refactor as each bump represents a missing abstraction. 3) The larger each bump – that is, the more lines of code it spans – the harder it is to build up a mental model of the function. The nesting depth for what is considered a bump is  levels of conditionals.

### How to fix it?
Bumpy Road implementations indicate a lack of encapsulation. Check out the detailed description of the [Bumpy Road code health issue](https://codescene.com/blog/bumpy-road-code-complexity-in-context/).
A Bumpy Road often suggests that the function/method does too many things. The first refactoring step is to identify the different possible responsibilities of the function. Consider extracting those responsibilities into smaller, cohesive, and well-named functions. The [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html) refactoring is the primary response.

**Comment by leynos** [comment]
@coderabbitai Re:

> I've already provided a detailed fix for this "Bumpy Road Ahead" issue in pump_stream_files earlier in this thread. Please see my previous response above where I suggested

If the remaining detection by CodeScene is a false positive, please provide a suitable suppression message to pass to codescene for this bumpy road

**Comment by leynos** [comment]
@coderabbitai review@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- On Windows, `_streams_rs._convert_fd_for_platform()` already converts a Python fd to an OS handle using `msvcrt.get_osfhandle`, but `convert_fd()` in Rust then treats the incoming value as an unsigned `u64`; since OS handles can be negative, this will raise `ValueError` for valid handles — consider aligning the Rust-side type/validation with the actual handle range returned by `get_osfhandle`.
- The helper utilities for pipe setup/teardown and reading (`_safe_close`, `_read_all`, etc.) are duplicated between `cuprum/unittests/test_rust_streams.py` and `tests/behaviour/test_rust_streams_behaviour.py`; consider extracting these into a shared test helper module to avoid divergence and make future changes simpler.

## Individual Comments

### Comment 1
<location> `rust/cuprum-rust/src/lib.rs:16-21` </location>
<code_context>
+#[cfg(unix)]
+use std::os::fd::FromRawFd;
+
+#[cfg(windows)]
+use std::os::windows::io::{FromRawHandle, RawHandle};
+
</code_context>

<issue_to_address>
**issue (bug_risk):** Potential failure on 64-bit Windows when HANDLE has the high bit set, due to signed i64 conversion.

On Windows `PlatformFd` is `usize`, but this function takes an `i64` and then calls `u64::try_from(value)`. Real OS handles on 64‑bit Windows can have the high bit set, so `get_osfhandle()` can return a negative `intptr_t` (and thus a negative Python `int`). In that case `u64::try_from(value)` will fail even though the handle is valid.

To avoid a platform bug here, consider treating the value as unsigned at the FFI boundary (e.g. accept `u64`) or reinterpreting negative `i64` values by their bit pattern instead of rejecting them.
</issue_to_address>

### Comment 2
<location> `rust/cuprum-rust/src/lib.rs:93-102` </location>
<code_context>
+    unsafe { File::from_raw_handle(handle as RawHandle) }
+}
+
+fn pump_stream(
+    reader_fd: PlatformFd,
+    writer_fd: PlatformFd,
+    buffer_size: usize,
+) -> Result<u64, io::Error> {
+    let mut reader = file_from_raw(reader_fd);
+    let mut writer = file_from_raw(writer_fd);
+    let result = pump_stream_files(&mut reader, &mut writer, buffer_size);
+    // The caller owns the reader FD; avoid closing it here.
+    std::mem::forget(reader);
+    result
+}
</code_context>

<issue_to_address>
**question (bug_risk):** File descriptor/handle ownership is asymmetric between reader and writer, which may be surprising to callers.

This converts both `reader_fd` and `writer_fd` into owned `File`s, but only the reader is `mem::forget`-ed, so the writer is closed on drop. The comment documents preserving the reader FD but says nothing about the writer, so the current behavior (auto-closing the writer) may surprise callers expecting both FDs to remain valid. If closing the writer/stdin is intended (e.g. to signal EOF), consider making that ownership transfer explicit and symmetric in both implementation and documentation (or naming), e.g. by also forgetting the writer or clearly treating it as consumed.
</issue_to_address>

### Comment 3
<location> `cuprum/_streams_rs.py:20-29` </location>
<code_context>
+    return importlib.import_module("cuprum._rust_backend_native")
+
+
+def _convert_fd_for_platform(fd: int) -> int:
+    """Convert a file descriptor for platform-specific Rust handling."""
+    if os.name != "nt":
+        return fd
+    import msvcrt
+
+    # Use getattr to avoid cross-platform stub mismatches in type checking.
+    get_osfhandle = typ.cast(
+        "typ.Callable[[int], int]",
+        getattr(msvcrt, "get_osfhandle"),  # noqa: B009
+    )
+    return get_osfhandle(fd)
+
+
</code_context>

<issue_to_address>
**issue (bug_risk):** Interaction between `get_osfhandle` and Rust-side signed conversion can reject valid Windows handles.

On Windows, `_convert_fd_for_platform` passes `msvcrt.get_osfhandle(fd)` into `native.rust_pump_stream` as an `i64`, but `get_osfhandle` returns an `intptr_t`, which can be negative on 64‑bit Windows when the high bit is set. The Rust side then does `u64::try_from(value)`, so any such negative (but valid) handle is rejected with `ValueError("file handle out of range")`. The boundary contract needs to be fixed: either normalize the handle to a non‑negative integer on the Python side (e.g., mask to `2**64 - 1` or similar) or update the Rust code to treat the incoming value as an unsigned handle instead of requiring a non‑negative `i64`.
</issue_to_address>

### Comment 4
<location> `cuprum/unittests/test_rust_streams.py:83-90` </location>
<code_context>
+    return output, transferred
+
+
+def test_rust_pump_stream_transfers_bytes() -> None:
+    """rust_pump_stream transfers bytes between pipes."""
+    streams = _load_streams_rs()
+    payload = b"cuprum-stream-payload"
+    output, transferred = _pump_payload(streams, payload)
+
+    assert output == payload
+    assert transferred == len(payload)
+
+
</code_context>

<issue_to_address>
**suggestion (testing):** Consider adding a test for empty/zero-byte input to validate EOF handling and return value 0.

We currently only cover non-empty payloads. Please add a test that writes no data (or closes the write end immediately) and asserts that `rust_pump_stream` returns 0 without raising, to cover the EOF / `read_len == 0` termination path.

```suggestion
def test_rust_pump_stream_transfers_bytes() -> None:
    """rust_pump_stream transfers bytes between pipes."""
    streams = _load_streams_rs()
    payload = b"cuprum-stream-payload"
    output, transferred = _pump_payload(streams, payload)

    assert output == payload
    assert transferred == len(payload)


def test_rust_pump_stream_transfers_zero_bytes() -> None:
    """rust_pump_stream handles zero-byte input and returns 0 (EOF path)."""
    streams = _load_streams_rs()
    payload = b""
    output, transferred = _pump_payload(streams, payload)

    assert output == payload
    assert transferred == 0
```
</issue_to_address>

### Comment 5
<location> `cuprum/unittests/test_rust_streams.py:130-139` </location>
<code_context>
+        _safe_close(write_fd)
+
+
+def test_rust_pump_stream_ignores_broken_pipe() -> None:
+    """rust_pump_stream drains input even if the writer breaks."""
+    streams = _load_streams_rs()
+    payload = b"broken-pipe"
+    in_read, in_write = os.pipe()
+    out_read, out_write = os.pipe()
+
+    try:
+        _safe_close(out_read)
+        os.write(in_write, payload)
+        _safe_close(in_write)
+
+        transferred = streams.rust_pump_stream(in_read, out_write)
+    finally:
+        _safe_close(in_read)
+        _safe_close(in_write)
+        _safe_close(out_read)
+        _safe_close(out_write)
+
+    assert isinstance(transferred, int)
+    assert transferred <= len(payload)
</code_context>

<issue_to_address>
**suggestion (testing):** Strengthen the broken-pipe test to assert that the upstream stream is fully drained and that no exception is raised.

The current assertions on return type and transfer count are useful, but you can more directly validate the "drain input" behavior by:
- using a larger payload (e.g., several buffer sizes),
- after `rust_pump_stream` returns, doing an extra `os.read(in_read, ...)` before closing `in_read` and asserting it returns `b""`, and
- explicitly asserting that no `OSError` is raised on BrokenPipe/ConnectionReset (e.g., `pytest.fail` in an `except` block).
This would more clearly prove that the reader side is fully consumed as documented.

Suggested implementation:

```python
def test_rust_pump_stream_ignores_broken_pipe() -> None:
    """rust_pump_stream drains input even if the writer breaks."""
    streams = _load_streams_rs()
    # Use a large payload to ensure we span multiple internal buffer reads.
    payload = b"x" * (64 * 1024)
    in_read, in_write = os.pipe()
    out_read, out_write = os.pipe()

    try:
        _safe_close(out_read)
        os.write(in_write, payload)
        _safe_close(in_write)

        try:
            transferred = streams.rust_pump_stream(in_read, out_write)
        except OSError as exc:
            # rust_pump_stream is expected to ignore broken pipe / connection reset
            # and drain the input instead of propagating the error.
            if getattr(exc, "errno", None) in (errno.EPIPE, errno.ECONNRESET):
                pytest.fail(
                    "rust_pump_stream raised OSError for broken pipe / connection reset"
                )
            raise

        # After rust_pump_stream returns, the input pipe should be fully drained.
        remaining = os.read(in_read, 4096)
        assert remaining == b""
    finally:
        _safe_close(in_read)
        _safe_close(in_write)
        _safe_close(out_read)
        _safe_close(out_write)

    assert isinstance(transferred, int)
    # All input bytes should have been consumed.
    assert transferred == len(payload)

```

To make this compile and run, ensure `errno` is imported at the top of `cuprum/unittests/test_rust_streams.py`, e.g.:

- If there is an `import os` line, extend it to `import errno, os`
  or add a separate `import errno` near the other imports.
</issue_to_address>

### Comment 6
<location> `docs/execplans/4-2-1-rust-pump-stream.md:155-156` </location>
<code_context>
+- Configurable buffer size behaviour (non-default value).
+- Error path when an invalid or closed file descriptor is supplied (expect
+  `OSError`).
+- Broken pipe handling: downstream closes early, function should not raise and
+  should continue draining input to match Python semantics.
+
+Add behavioural tests using pytest-bdd that exercise the Rust pump stream when
</code_context>

<issue_to_address>
**nitpick (typo):** Tighten grammar in the broken pipe handling description.

"downstream closes early, function should not raise" is a comma splice; consider "downstream closes early; the function should not raise" or a similar rephrasing.

```suggestion
- Broken pipe handling: downstream closes early; the function should not raise
  and should continue draining input to match Python semantics.
```
</issue_to_address>

### Comment 7
<location> `rust/cuprum-rust/src/lib.rs:143` </location>
<code_context>
+            )
+        })?;
+
+        match handle_write(writer, chunk) {
+            Ok(bytes) => total_written = total_written.saturating_add(bytes),
+            Err(err)
</code_context>

<issue_to_address>
**suggestion (review_instructions):** This `match` has three branches and should be extracted into a predicate/helper to keep multi-branch conditionals out of the main control flow.

The `match handle_write(writer, chunk)` has three distinct branches (`Ok`, `Err` for broken pipe / reset, and the general `Err`). Per the guideline to move conditionals with more than two branches into a predicate function, consider extracting this into a small helper (e.g. `fn handle_write_result(...) -> Result<(u64, bool), io::Error>` or similar) so that the loop body stays simpler and the multi-branch logic is encapsulated.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>

### Comment 8
<location> `docs/execplans/4-2-1-rust-pump-stream.md:14` </location>
<code_context>
+## Purpose / big picture
+
+Provide a Rust-backed `rust_pump_stream()` that transfers data between file
+descriptors outside the GIL with a configurable buffer (default 64 KB) and
+clean error propagation to Python exceptions. The new function must be fully
+covered by unit and behavioural tests, and documentation must reflect the Rust
</code_context>

<issue_to_address>
**issue (review_instructions):** The acronym “GIL” is used here without being expanded on first use; please spell it out (for example, Global Interpreter Lock) the first time it appears in this document.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 9
<location> `docs/execplans/4-2-1-rust-pump-stream.md:53` </location>
<code_context>
+
+- Risk: Rust file descriptor handling may accidentally close file descriptors
+  owned by Python, causing subtle downstream failures. Severity: high
+  Likelihood: medium Mitigation: avoid closing the reader FD by forgetting the
+  `File` wrapper. Allow the writer FD to close when the pump completes.
+- Risk: GIL release might not cover the full I/O loop, reducing throughput.
</code_context>

<issue_to_address>
**issue (review_instructions):** The acronym “FD” is introduced here without definition; please expand it (for example, file descriptor) on first use in the document.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: cuprum/_streams_rs.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    if os.name != "nt":
+        return fd
+    import ctypes
+    import msvcrt
+
+    # Use getattr to avoid cross-platform stub mismatches in type checking.
+    get_osfhandle = typ.cast(
+        "typ.Callable[[int], int]",
+        getattr(msvcrt, "get_osfhandle"),  # noqa: B009
+    )
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/_streams_rs.py:39]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[cuprum/unittests/test_rust_streams.py](https://github.com/leynos/cuprum/pull/29/files/ca0b947d13db4ae409bff386643b934d7cb2cb20#diff-3708370c7648b8a254b1f926a193dc609f13fb6f795978eae585e97ad79745d0)

Comment on lines +51 to +67

~~~python
def test_rust_pump_stream_transfers_bytes(rust_streams: ModuleType) -> None:
    """Validate that rust_pump_stream transfers bytes between pipes.

    Parameters
    ----------
    rust_streams : ModuleType
        The Rust streams module fixture.

    Returns
    -------
    None
    """
    payload = b"cuprum-stream-payload"
    output, transferred = _pump_payload(rust_streams, payload)

    assert output == payload, "expected payload to round-trip through pump"
    assert transferred == len(payload), "expected transferred count to match payload"
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=29&biomarker=Code+Duplication&filename=cuprum%2Funittests%2Ftest_rust_streams.py&method=)
The module contains 3 functions with similar structure: test_rust_pump_stream_respects_buffer_size,test_rust_pump_stream_transfers_bytes,test_rust_pump_stream_transfers_zero_bytes

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[cuprum/unittests/test_rust_streams.py](https://github.com/leynos/cuprum/pull/29/files/ca0b947d13db4ae409bff386643b934d7cb2cb20#diff-3708370c7648b8a254b1f926a193dc609f13fb6f795978eae585e97ad79745d0)

Comment on lines +51 to +67

~~~python
def test_rust_pump_stream_transfers_bytes(rust_streams: ModuleType) -> None:
    """Validate that rust_pump_stream transfers bytes between pipes.

    Parameters
    ----------
    rust_streams : ModuleType
        The Rust streams module fixture.

    Returns
    -------
    None
    """
    payload = b"cuprum-stream-payload"
    output, transferred = _pump_payload(rust_streams, payload)

    assert output == payload, "expected payload to round-trip through pump"
    assert transferred == len(payload), "expected transferred count to match payload"
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=29&biomarker=Code+Duplication&filename=cuprum%2Funittests%2Ftest_rust_streams.py&method=)
The module contains 3 functions with similar structure: test_rust_pump_stream_respects_buffer_size,test_rust_pump_stream_transfers_bytes,test_rust_pump_stream_transfers_zero_bytes
